File: 76-minimum-window-substring/minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if (t.empty()) return "";

        unordered_map<char, int> tmap;
        unordered_map<char, int> smap;

        for (char c : t) {
            tmap[c]++;
        }

        int have = 0, need = tmap.size();
        int i = 0;
        int resLen = INT_MAX;
        pair<int, int> res;

        for(int j = 0; j < s.size(); j++) {
            smap[s[j]]++;

            if (tmap.contains(s[j]) && smap[s[j]] == tmap[s[j]])
                have++;

            while(have == need) {
                if ((j - i + 1) < resLen) {
                    resLen = j - i + 1;
                    res = {i, j};
                }
                smap[s[i]]--;
                if (tmap.contains(s[i]) && smap[s[i]] < tmap[s[i]])
                    have--;
                
                i++;
            }
        }

        if (resLen == INT_MAX)
            return "";

        return s.substr(res.first, resLen);
    }
};

File: 76-minimum-window-substring/minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(t) == 0:
            return ""

        smap = {}
        tmap = Counter(t)
        length = sys.maxsize
        index = 0

        have, need = 0, len(tmap)

        i = 0
        for j in range(len(s)):
            smap[s[j]] = 1 + smap.get(s[j], 0)

            if (s[j] in tmap and smap[s[j]] == tmap[s[j]]):
                have += 1
            
            while have == need:
                if (j-i+1) < length:
                    length = j-i+1
                    index = i
                
                smap[s[i]] -= 1
                
                if s[i] in tmap and smap[s[i]] < tmap[s[i]]:
                    have -= 1
                
                i += 1
        
        if length == sys.maxsize:
            return ""
        
        return s[index: index+length]



